### DISCUSSION TITLE
How to invoke local serverless function from within a function

### DISCUSSION THREAD

**Discussion by modsushi** [vercel/community#1473] [Help]
I want to trigger another function from within a function. Can't find any documentation on this.

**Comment by amyegan** [comment]
Hi, Jabran! I think you're looking for something like the [examples/edge-functions/api-rate-limit-and-tokens/components/api-request.tsx#L4](https://github.com/vercel/examples/blob/dabe58d6bc72a752ef3ae013abc561809c29228a/edge-functions/api-rate-limit-and-tokens/components/api-request.tsx#L4). It shows how to fetch data from a serverless function.

If I misunderstood, can you explain a little bit more about your scenario?

**Comment by modsushi** [comment]
I have 2 serverless functions. One is using JavaScript while the other is on Python runtime. During execution the JavaScript function makes a HTTP request to the python function (invoking it) by using the full http address (mydomain/api/fn).
There must be a way to internally call the other function using local routing.

I am guessing this is a pretty normal usecase of calling another function from a function. Since AWS lambdas have specific documentation on this.

**Comment by frencojobs** [comment]
> There must be a way to internally call the other function using local routing

Uh nope. Don't think so. I'm guessing you mean relative paths by "local routing". You need an absolute URL to perform an HTTP request. Except in browsers, because both `XMLHttpRequest` and `fetch` will induce the the absolute URL from the relative path by design, which is possible because a web page know it's own URL. Servers, well functions in this case, don't know where they'll end up on when you run them, that is until your request reaches them.

Once the request reaches to the function, you can induce the absolute URL using `req.headers.host` and `req.url`.
```js
// for some.vercel.app/api/foo
// req.headers.host -> some.vercel.app
// req.url          -> /api/foo
```
**But that is a terrible idea**.

<details>
<summary>Here's why.</summary>

Let's say you have two functions, `/api/foo` and `/api/bar`. And let's say `/api/foo` sends passwords to `/api/bar` like this.
```js
const bar_endpoint = `https://${req.headers.host}/api/bar`;
send_password_to(bar_endpoint);
```
Then, a malicious person could easily create a proxy server of some sort (let's say `steal.vercel.app`) and point `/api/foo` to your server and `/api/bar` to the malicious one. Here's an example using [Vercel Rewrites](https://vercel.com/docs/concepts/projects/project-configuration#rewrites),
```json5
// vercel.json
{
  "rewrites": [
    {
      "source": "/api/foo",
      "destination": "https://some.vercel.app/api/foo"
    },
    {
      "source": "/api/bar",
      "destination": "https://steal.vercel.app/steal-passwords-mwah-ha-ha"
    }
  ]
}

```
Because `req.headers.host` is based on the request, `/api/foo` will receive it as `steal.vercel.app` instead of `some.vercel.app`. And suddenly your passwords are gone.

</details>

The gist is you need the full URL to communicate (aka perform an HTTP request) between functions (which are effectively HTTP endpoints).

> AWS lambdas have specific documentation on this

Not sure what you mean by this. Would be great if you can share a link. We have had use cases like yours and we always have to specify the endpoints as Environment Variables through Parameter Store.

**Comment by adrientiburce** [comment]
Hello,

I'm asking myself the same question. 

I got 2 endpoints : 
- /api/questions:  an edge function 
- /api/data:  a serverless function

I want to call `/api/data` from `/api/questions`, and I'm doing this because : 
- I'm using a lib in /api/data that is not working with edge functions

How can I invoke the base url (my code is pretty ugly IMO): 

```js 
// api/questions.js file

let baseURL = "http://localhost:3000"
if (process.env.API_URL) {
  baseURL = process.env.API_URL
}


async function fetchData(): Promise<any> {
  const response = await fetch(`${baseURL}/api/data`);
  const data = await response.json();

  return data
}
```

Thanks !

**Reply by ashubham** [reply]
I feel this request makes sense. I have the same use case with multiple serverless functions. Now when I deploy previews etc, how do I test them together ?

**Reply by amyegan** [reply]
Since the functions seem to be in the same project, there are a few options involving preset environment variables. There is a `VERCEL_URL` environment variable, but it is set to the generated deployment URL without `https://`. Or you could use a ternary with `VERCEL_ENV` and your known custom domain, like
```
baseURL = process.env.VERCEL_ENV === "production" ? "example.com" : process.env.VERCEL_URL
```

**Reply by modsushi** [reply]
you'd need to enable CORS on the function to call it from serverless function.
Would need some work on securing the endpoint
